Add Taproot support

dgpv opened the pull request:
This adds a bunch of new APIs and it is really important for me to hear opinions on the convenience/ergonomics of these APIs. Please don't hesitate to comment if you think something might be made more convenient for the developer.

Of course the correctness/safety is the most important, but convenience is close to that in importance, as it can reduce the chance for mistakes

dgpv commented on bitcointx/core/bitcoinconsensus.py:
```diff
@@ -68,20 +69,16 @@
-    bitcoinconsensus_SCRIPT_FLAGS_VERIFY_NULLDUMMY | bitcoinconsensus_SCRIPT_FLAGS_VERIFY_CHECKLOCKTIMEVERIFY |
-    bitcoinconsensus_SCRIPT_FLAGS_VERIFY_CHECKSEQUENCEVERIFY | bitcoinconsensus_SCRIPT_FLAGS_VERIFY_WITNESS
-)
+bitcoinconsensus_SCRIPT_FLAGS_VERIFY_TAPROOT = 1 << 17
```
Taproot is not currently supported in libbitcoinconsensus, but I used an ad-hoc patch that adds such support (https://gist.github.com/dgpv/7bf8787ecb093e12c862287f4b86c3f4, not suitable for PR to Core, at all) to test taproot script with libbitcoinconsensus. Currently taproot support for ConsensusVerifyScript is disabled. When the support will be available in libbitcoinconsensus in Core, it will be easy to adjust this code in python-bitcointx and re-enable it.

AdamISZ replied:
Yes, I see, it's a bit hairy; I remember that you have detailed explanation on your readme about this `VerifyScript` vs `ConsensusVerifyScript` point. It isn't going to be a big deal for our kinds of future use cases, the more imperfect should be enough as we either (a) don't have a risk from script invalidity, (b) are using very vanilla scripts anyway. But presumably libbitcoinconsensus will get this taproot update at some point. So:

> Currently taproot support for ConsensusVerifyScript is disabled. 

seems entirely logical, does it mean that it *is* supported for VerifyScript, though?

dgpv replied:
No, it is not supported for VerifyScript. The python-code interpreter is not something that I feel the energy to support in up-to-date state (there is quite a lot of things not implemented for VerifyScript even not considering taproot). I'm not removing the interpreter code, though - maybe someone in the future will make a PR to bring it more in-sync with Core...

AdamISZ replied:
OK that's fine, understood, thanks for clarifying.

dgpv commented on bitcointx/core/key.py:
```diff
@@ -53,6 +55,13 @@
+    ...
+
+
+NO_MERKLE_ROOT_TWEAK = _flag_NO_MERKLE_ROOT_TWEAK()
```
Because in the most cases the key is tweaked when signing, I decided to use a special 'no tweak' flag instead of relying on distinction of `merkle_root` being "empty bytes" or "None".

"empty bytes"  `merkle_root` means that there is a tweak, that is derived from the pubkey.

AdamISZ replied:
Indeed, that seems to agree with the recommendation of BIP341.

dgpv commented on bitcointx/core/key.py:
```diff
@@ -1905,6 +2030,162 @@ def get_pubkey(self, key_id: bytes,
+        return True
+
+    def create_tap_tweak(self: T_XOnlyPubKey, *,
+                         merkle_root: bytes = b'',
```
requiring `merkle_root` to always be specified as kwarg may feel a bit excessive for the method with just one argument, but I think this is OK since a lot of other functions/methods use `merkle_root` kwarg, when there are other args, and this way it feels more uniform

dgpv commented on bitcointx/core/script.py:
```diff
@@ -1693,6 +1887,264 @@ def collect_sig(self, pub: 'bitcointx.core.key.CPubKey',
+            get_script_with_control_block() method to be available
+
+        'leaf_version' can be supplied for all the script in the tree to
+            have this version
```
I decided to not add `leaf_version` field to `CScript`, at least for now. It is not clear to me if the single different leaf version for one script among other scripts will be a normal occurence, even when we'll have many leaf versions. It seems more possible that the whole trees will consist of scripts with particular leaf version. And you can always have a 'tree' of just one script to designate a leaf version for just that script.

dgpv commented on bitcointx/core/serialize.py:
```diff
@@ -426,22 +426,6 @@ def stream_deserialize(cls, f: ByteStream_Type, **kwargs: Any) -> List[int]:
         return ints
 
 
-class VarBytesSerializer(Serializer[bytes]):
```
This was a duplicate of `BytesSerializer`

dgpv commented on bitcointx/wallet.py:
```diff
@@ -476,6 +483,142 @@ def from_redeemScript(cls: Type[T_P2WPKHCoinAddress],
+        pubkey: Union[XOnlyPubKey, bytes, bytearray],
+        *,
+        accept_invalid: bool = False
+    ) -> T_P2TRCoinAddress:
```
There is a distinction with `from_*output_pubkey()` because by default the address will be created from the internal pubkey, with a tweak derived from that same pubkey. If the tweak was created manually, then one will use `from_*output_pubkey()`

AdamISZ replied:
Is there a typo there? I think you meant `from_xonly_pubkey` the second time? Perhaps change the latter to `from_xonly_internal_pubkey` (aligning with 'taproot internal key' in the BIP).

dgpv replied:
I've fixed the previous message to make it clear.

I'm not sure that `from_output_pubkey()` is even useful - if someone has create a tweaked pubkey, it certainly will be `XOnlyPubkey`. Why would they then convert it to `CPubKey` that they can supply to `from_output_pubkey()` ?

But if we make ``from_output_pubkey()` to accept both `CPubKey` and `XOnlyPubKey`, then at least we can say that the function name is shorter without the `_xonly_` part...

AdamISZ replied:
Hmm yes, sometimes options/convenience functions can even end up making things more confusing. But before I expand on that, I'd say: this is pretty trivial don't worry about it. If there isn't some massive footgun lurking, then the details of what convenience functions you expose don't matter that much.

So I'd say, as a consumer of the library, I'd expect it to offer one function for CPubkey->XonlyPubkey (which tbh i think should be better called BIP340Pubkey, because it has an implicit relation to private keys that's specific to BIP340), and all the actual taproot functionality just goes from there, i.e. I expected the pubkey-to-address function for taproot addresses to consume only BIP340 pubkeys.

But just because "I'd expect that" doesn't really mean too much, if you have extra functions I'm not bothered at all.

dgpv replied:
> which tbh i think should be better called BIP340Pubkey

I try to follow Core source where it does not hinder safety or ergonomics, `XOnlyPubKey` is from Core.

dgpv commented on bitcointx/wallet.py:
```diff
@@ -476,6 +483,142 @@ def from_redeemScript(cls: Type[T_P2WPKHCoinAddress],
+    @classmethod
+    def from_pubkey(cls: Type[T_P2TRCoinAddress],
+                    pubkey: Union[CPubKey, bytes, bytearray],
+                    ) -> T_P2TRCoinAddress:
```
It is not clear to me if we should allow to supply `XOnlyPubkey` to `from_pubkey()` and `from_output_pubkey()`. We could just check for the type and call the equivalent `from_xonly_*` method. But maybe the strictness is better here ? The wrong type of pubkey could be catched earlier in some code...

I really want to hear opinions from others on this.

AdamISZ replied:
Pretty much as above, I'd have expected only one way to make these addresses, i.e. from x-only pubkeys, but I have no problem with you adding more ways to do it. One thing I would ask though, is that the docstrings are clearer; as per some earlier conversation, I found `*output*` not immediately obvious, i almost feel like the function name `from_internal_key_without_script` or something like that, would be the most obvious, but in any case as a user I'd need to read the docstring to actually be sure of what the function is doing (even if I read the BIP, I'd need to know exactly how you chose to follow the guidelines).

dgpv replied:
OK, noted, naming and docstrings.

I recently wrote some sample code that generates P2TR addresses, and found out that I instinctively wrote `P2TRCoinAddress.from_pubkey(internal_pub)`, while `internal_pub` was `XOnlyPubKey`. This gave an error because `from_pubkey()` only takes `CPubKey`. This makes me think that allowing `XOnlyPubKey` for `P2TRCoinAddress.from_pubkey` is actually not a bad idea. The amount of possible incorrect behaviour the restriction would catch seems small to me, while the ergonomics improvement seems to be big enough to justify allowing `from_pubkey` to take `XOnlyPubKey`, too.

If there is no objections, I'll do it.

dgpv replied:
I've enabled `P2TRCoinAddress.from_pubkey` to take x-only pubkeys, and I updated the docstrings for `P2TRCoinAddress.from_*pubkey` methods: https://github.com/Simplexum/python-bitcointx/commit/3dfd1c64e568c8814a43a2066be5c409beaa91aa

I decided to stick with `from_xonly_pubkey` and `from_xonly_output_pubkey` naming, I hope that the reference to BIP341 in the docstring will be enough for clearer understanding by the users of the lib.

dgpv commented:
This adds a bunch of new APIs and it is really important for me to hear opinions on the convenience/ergonomics of these APIs. Please don't hesitate to comment if you thing something might be made more convenient for the developer.

Of course the correctness/safety is the most important, but convenience is close to that in importance, as it can reduce the chance for mistakes

AdamISZ commented on bitcointx/core/key.py:
```diff
@@ -283,12 +311,101 @@ def sign_compact(self, hash: Union[bytes, bytearray]) -> Tuple[bytes, int]: # py
 
         return output.raw, recid.value
 
+    def sign_schnorr(
```
This might be a structural misunderstanding, but might not we want at some point to do Schnorr sigs with a `CKey` that are not taproot signings? Just quibbling therefore about the function name, i.e. this seems to be implementing BIP341 over BIP340, not only BIP340.

dgpv replied:
The function name is just copied from Core's `CKey::SignSchnorr`. The difference bip341/bip340, IIUC, is only in ability to supply `merkle_root`. To sign without tweak, just supply `NO_MERKLE_ROOT_TWEAK` as `merkle_root` arg.

`test_schnorr()` in `tests/test_key.py` does that, and checks the signing against the test vectors from BIP340.

AdamISZ replied:
oh i didn't spot that `NO_MERKLE_ROOT_TWEAK` is literally what is says, as distinct from the "null" tweak of H(P); so since this bypasses any taproot processing that's all good.

dgpv replied:
Ah, maybe I understood that you meant re naming. The default behavior of `sign_schnorr()` is to sign with a tweak because `merkle_root` is by default `b''`. With Core's `CKey::SignSchnorr` there's no 'default' value, you supply `merkle_root` as null or as uint256 instance. Equivalent in python would be making `merkle_root` as a non-kwarg argument, but that does not seem too ergonomic to me.

We could make `sign_schnorr` to sign without tweak by default, but that means that when someone wants to sign for a simple taproot non-script spend, they would need to do `k.sign_schnorr(merkle_root=b'')` istead of just `k.sign_schnorr()`.

My reason for doing it with tweak by default is that it seemed that no-tweak signing will likely be used more rarely than 'tweaked' signing, and people can just bear with supplying `NO_MERKLE_ROOT_TWEAK` explicitly in that case.

We could drop `sign_schnorr()` and use two funcs `sign_bip340()` and `sign_bip341()`, but i like `sign_schnorr()` better, and it corresponds to the function name in Core.

AdamISZ replied:
yeah i'm fine with tweak-exists-by-default, I don't see anything bad about that. at some point it's just preferences, and there's nothing unreasonable here.

dgpv replied:
Hmm, could `NO_MERKLE_ROOT_TWEAK` be read/misunderstood as "no merkle root"-tweak ? I.e. 'a tweak that is applied when you have no (empty) merkle root, instead of ``no "merkle_root" tweak'' ?

I can't come up with better name, though...

dgpv replied:
`SKIP_MERKLE_ROOT_TWEAK` ?

AdamISZ replied:
`NO_MERKLE_ROOT_TWEAK` is fine I think. Just an error on my part.

AdamISZ commented on bitcointx/__init__.py:
```diff
@@ -310,6 +310,33 @@ def select_chain_params(params: Union[str, ChainParamsBase,
+    currently marked as 'experimental'. These modules being experimental
+    mean that their ABI can change at any moment. Therefore, to use
+    python-bitcointx with this functionality, it is highly recommended to
+    compile your own version libsecp256k1 using the git commit that is
```
If I want to test the PR now, which git commit should I use?

dgpv replied:
7006f1b97fd8dbf4ef75771dd7c15185811c3f50 (this is mentioned in release-notes.md)

AdamISZ replied:
Thanks.

AdamISZ replied:
So I built with `./configure --enable-experimental --enable-module-schnorrsig --enable-module-ecdh`, on that commit, and the test suite is passing except for the `libbitcoinconsensus` part, as expected. But I'm a little confused because the schnorrsig module is separate from the experimental part, no? (I guess I need to read the codebase to understand what exact functionality is currently 'experimental' here).

AdamISZ replied:
Oh I see; it is explicity here:

https://github.com/Simplexum/python-bitcointx/pull/60/files#diff-35298aa2172185ac9b01727f6963d68e59c79d6ce7a02d9c3d4ee71ef29b06a5R80

dgpv replied:
Taproot-related script tests are disabled entirely, so they are not the cause of these tests failing, they failed just because libbitcoinconsensus library was not available.

libbitcoinconsensus tests should have been skipped rather than failing. I pushed fixed version that skips the tests if the lib is not available.

dgpv replied:
>  the schnorrsig module is separate from the experimental

scnorrsig is 'experimental' (along with 'extrakeys', but schnorrsig depends on extrakeys, so you actually do'nt need to do `--with-module-extrakeys` when you have `--with-module-schnorrsig`)

if you omit `--enable-experimental` when configuring with  `--with-module-schnorrsig`, configure will fail.

AdamISZ replied:
> libbitcoinconsensus tests should have been skipped rather than failing in this case. I pushed fixed version that skips the tests if the lib is not available.

OK, I believe that works here, I have 5 tests skipped and the rest passed with `python -m unittest discover`:

```
Ran 159 tests in 4.671s

OK (skipped=5)
```

dgpv replied:
> 5 tests skipped

3 of those are likely skipped because you did not set `PYTHON_BITCOINTX_ALLOW_LIBSECP256K1_EXPERIMENTAL_MODULES_USE=1` and there was no attempt to use schnorrsig functions at all

AdamISZ replied:
Hmm still something off; I did `PYTHON_BITCOINTX_ALLOW_LIBSECP256K1_EXPERIMENTAL_MODULES_USE=1 python -m unittest discover` and now *6* tests are being skipped, not 2 :) On `d776833a9b7a83eb47fce0c4e18a0d6640f64896`. Here is the output:

```
............................................................................/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:159: UserWarning: libbitcoinconsensus library is not avaliable, not testing bitcoinconsensus module and taproot scripts
  warnings.warn(
./home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:159: UserWarning: libbitcoinconsensus library is not avaliable, not testing bitcoinconsensus module and taproot scripts
  warnings.warn(
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 2 tests: CHECKSEQUENCEVERIFY not supported
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 1 tests: CHECKSEQUENCEVERIFY not supported
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 39 tests: MINIMALDATA not supported
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 2 tests: SIG_PUSHONLY check not implemented
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 75 tests: MINIMALDATA check not implemented
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 1 tests: SIG_PUSHONLY check not implemented
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 3 tests: SIG_PUSHONLY check not implemented
  warnings.warn("SKIPPED {} tests: {}"
/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:103: UserWarning: SKIPPED 5 tests: CHECKSEQUENCEVERIFY not supported
  warnings.warn("SKIPPED {} tests: {}"
./home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:159: UserWarning: libbitcoinconsensus library is not avaliable, not testing bitcoinconsensus module and taproot scripts
  warnings.warn(
s/home/waxwing/code/python-bitcointx/bitcointx/tests/test_scripteval.py:159: UserWarning: libbitcoinconsensus library is not avaliable, not testing bitcoinconsensus module and taproot scripts
  warnings.warn(
s.............ssss..............................................................
----------------------------------------------------------------------
Ran 159 tests in 4.903s

OK (skipped=6)
```

dgpv replied:
Can you do `PYTHON_BITCOINTX_ALLOW_LIBSECP256K1_EXPERIMENTAL_MODULES_USE=1 python -m unittest discover --verbose 2>&1|grep skipped` so we can see exactly what tests were skipped and why ?

AdamISZ replied:
Oh that seems to have helped quite a bit :)

```
skipped 'bitcoinconsensus library is not available'
skipped 'bitcoinconsensus library is not available'
test_sign_message_simple (bitcointx.tests.test_signmessage.Test_SignVerifyMessage) ... skipped 'secp256k1 compiled without pubkey recovery functions'
test_sign_message_vectors (bitcointx.tests.test_signmessage.Test_SignVerifyMessage) ... skipped 'secp256k1 compiled without pubkey recovery functions'
test_verify_message_simple (bitcointx.tests.test_signmessage.Test_SignVerifyMessage) ... skipped 'secp256k1 compiled without pubkey recovery functions'
test_verify_message_vectors (bitcointx.tests.test_signmessage.Test_SignVerifyMessage) ... skipped 'secp256k1 compiled without pubkey recovery functions'
OK (skipped=6)
```
so i missed `--enable-module-recovery` i guess

AdamISZ replied:
Yep that fixed it. Now only skips the 2 tests.

AdamISZ commented on bitcointx/core/key.py:
```diff
@@ -283,12 +311,101 @@ def sign_compact(self, hash: Union[bytes, bytearray]) -> Tuple[bytes, int]: # py
+    ) -> bytes:
+
+        ensure_isinstance(hash, (bytes, bytearray), 'hash')
+        if len(hash) != 32:
```
maybe better not use `hash` since it's a builtin?

dgpv replied:
This is 'historical' in a sense that other functions in this file also use the name `hash`, this was in python-bitcoinlib code. I used `hash` for uniformity, but maybe I should have renamed all the `hash` vars thoughout the file instead ?

AdamISZ replied:
That's fine don't worry. Was a very minor comment, there is no need to create extra work for it.

AdamISZ commented on bitcointx/core/key.py:
```diff
@@ -1905,6 +2030,162 @@ def get_pubkey(self, key_id: bytes,
+class XOnlyPubKey(bytes):
+    """An encapsulated X-Only public key"""
+
+    __fullyvalid: bool
```
I realise that this is not specific to the new taproot/BIP340 code (you already had it in `CPubKey`, but what is the significance of storing this `__fullyvalid` bool? Is it somehow useful to have a pubkey object that is from an invalid encoding?

dgpv replied:
- The Core makes that distinction for some reason
- It is used in tests to supply invalid pubkeys and check that the code fail with invalid pubkeys

My guess is that Core does this for speed so you can create `CPubKey` instance and check for validity later.

Maybe it makes sense to throw `ValueError` on invalid encoding and introduce `allow_invalid_encoding` kwarg for the tests ? This would be breakage of the API, but I doubt this breakage will affect much real-world code except the test code.

dgpv replied:
Address classes already have `accept_invalid` kwarg for `from_pubkey()` method, but I guess one could accidentally use invalid pubkeys in other places (like, constructing scripts) where they could miss checking the `is_fullyvalid()`.

dgpv replied:
OTOH, if we introduce that change, and then new code that uses bitcointx starts dropping `is_fullyvalid()` checks, but afterwards somehow is used with the old version of the lib, then it could pass invalid pubkeys where it should not. This is somewhat alleviated by the fact that the code may list the version requirement for the library, of course.

AdamISZ replied:
I think only two reasonable paths: leave as is, or completely remove, anything else would have the kind of problems you describe in that last comment. Even if I think the 'remove' option is preferable, the taproot patch is not really the right time to make such a big change.

> Address classes already have accept_invalid kwarg for from_pubkey() method

It's interesting that this option exists (even though of course default False). I suppose it might be related to burn addresses? (your performance hypothesis makes sense too, i guess). Generally I'd lean towards making no public API support for such things, since the risk of using them incorrectly outweighs the very small  amount of usage (burn addresses are very rarely used for anything).

But no action item here, imo: this is orthogonal to the taproot patch.

dgpv replied:
and that `test_from_invalid_pubkeys` was present in python-bitcoinlib since long time ago. So maybe this is really just an option to allow to create burn addresses.

dgpv replied:
> But no action item here, imo: this is orthogonal to the taproot patch.

I created a separate issue: https://github.com/Simplexum/python-bitcointx/issues/61

AdamISZ commented on bitcointx/core/key.py:
```diff
@@ -1905,6 +2030,162 @@ def get_pubkey(self, key_id: bytes,
+        elif len(keydata) == 33:
+            ensure_isinstance(keydata, CPubKey,
+                              'x-only pubkey data of 33 bytes')
+            keydata = keydata[1:33]
```
Maybe I missed something obvious, but: why would we create an `XOnlyPubkey` from just a removal of the parity byte of a `CPubKey`? I get it that this might be fine in some contexts, but isn't there the danger that the caller is dealing with an existing private/pubkey pair that is not actually valid for XOnly, because it has the wrong parity?

dgpv replied:
From BIP340:

> the standard 33-byte compressed public key format consists of a byte indicating the oddness of the Y coordinate, plus the full X coordinate.

so we just drop the parity byte.

The code from Core:

```
    /** Construct an x-only pubkey from a normal pubkey. */
    explicit XOnlyPubKey(const CPubKey& pubkey) : XOnlyPubKey(Span<const unsigned char>(pubkey.begin() + 1, pubkey.begin() + 33)) {}
```

AdamISZ replied:
My concern is the scenario: I have a pre-existing key and for whatever reason I want to make "the taproot address for the same key", I think this is actually wrong-headed, but if you do it, you have some original privkey k, whose public key has coordinates x_k, y_k, if y_k is odd then when you change the pubkey from `03deadbeef..` to `deadbeef` you are actually encoding the pubkey corresponding to `-k`, not `k` (true because on the curve P + -P = O), so a person ignorantly doing this will find themselves unable to sign against this pubkey (although they have the trivial solution of flipping the sign of their privkey, admittedly).

AdamISZ replied:
OK I guess this isn't really anything to discuss, it's already the case that the signing algorithm is 'take private key and if it's not an even y, then flip its sign', so this is already addressed/understood.

dgpv replied:
There's another potential problem here to discuss: data taken from invalid `CPubKey` can actually constitute a valid `XOnlyPubKey`:

```
>>> CPubKey(x('0478d430274f8c5ec1321338151e9f27f4c676a008bdf8638d07c0b6be9ab35c71')).is_fullyvalid()
False
>>> XOnlyPubKey(x('78d430274f8c5ec1321338151e9f27f4c676a008bdf8638d07c0b6be9ab35c71')).is_fullyvalid()
True
```

I think the safe route would be to not accept invalid `CPubKey` here, those who really need to create XOnlyPubKey from invalid `CPubKey` can simply strip the first byte themselves.

AdamISZ replied:
Right. But I mean coming back to my first comment, I was only concerned that you would get "the wrong" key from parity stripping, but then as per convo I realized that concern doesn't make sense, because of the way signing works.

But what you raise here, i.e. that someone can start with a completely invalid key and after conversion, end with a valid one ... I mean, I honestly don't want to try to figure out the situations where that could cause a problem. The whole thing just feels kind of wrong for cryptographically or financially sensitive code: if you ever know that input is not correct (not fully valid), stop there, don't leave it around, creating a 'maybe valid in future' object.

I do get that this is coming directly from something in Core, but the KISS principle seems to say 'never create pubkey objects from invalid pubkey serializations'.

So in short I agree with your final sentence :)

dgpv replied:
I went and made the change to not accept invalid `CPubKey` as arguments to `XOnlyPubKey()`: https://github.com/Simplexum/python-bitcointx/commit/01f5b3f342604f213e9ce7575e42749514d21b5e

AdamISZ commented on bitcointx/core/key.py:
```diff
@@ -283,12 +311,103 @@ def sign_compact(self, hash: Union[bytes, bytearray]) -> Tuple[bytes, int]: # py
+        # but still, do this to be close to the original source
+        ctypes.memset(keypair_buf, 0, 64)
+
+        return sig_buf.raw
```
https://github.com/bitcoin-core/secp256k1/blob/7006f1b97fd8dbf4ef75771dd7c15185811c3f50/include/secp256k1_schnorrsig.h#L97

... the comments point out that using `schnorrsig_sign` does not automatically verify the created signature. Maybe we should have a built-in verify check, to agree with the verification check recommendation in BIP340?

dgpv replied:
Added the sig verification right after signing in https://github.com/Simplexum/python-bitcointx/commit/d776833a9b7a83eb47fce0c4e18a0d6640f64896

AdamISZ commented on bitcointx/core/script.py:
```diff
@@ -1364,6 +1440,128 @@ def SignatureHash(script: CScript, txTo: 'bitcointx.core.CTransaction', inIdx: i
+    # Hash type
+    if hashtype is None:
+        hashtype = SIGHASH_ALL
+        hashtype_byte = b'\x00'
```
Wait, did I get something wrong here? I thought SIGHASH_ALL was always 0x01, and BIP341 explicitly defines SIGHASH_DEFAULT as 0x00 (although they both have the same semantics, which is weird - I remember explicitly asking sipa about this back during the taproot peer review sessions, as the original wording in the BIP was confusing to those who didn't know this obscure fact about the 0x00 sighash flag - afaict *all* code/wallets had always been using 0x01). But I could be misremembering something, quite easily.

dgpv replied:
Here, `hashtype is None` has the same meaning as `SIGHASH_DEFAULT` in Core. It is the same as `SIGHASH_ALL` in all regards except that the sighash data to be hashed will have 0x00 instead of 0x01. Because of this, I introduced `sighash_byte`.

I did not want to add `SIGHASH_DEFAULT` to `SIGHASH_Type`, because it is used only for taproot and is essentially same with SIGHASH_ALL, so I wanted to avoid confusion and decidet to not introduce it at all.

AdamISZ replied:
OK, after a more careful read of BIP341 I see that this is the right way to do this part, since `SIGHASH_DEFAULT` is implicit for 64 bytes, which of course in the overwhelming majority of cases will be the one people use.

dgpv commented on bitcointx/core/key.py:
```diff
@@ -283,12 +311,124 @@ def sign_compact(self, hash: Union[bytes, bytearray]) -> Tuple[bytes, int]: # py
+        # Not that it matters much in python where we don't have control over
+        # memory and the keydata is probably spread all over the place anyway,
+        # but still, do this to be close to the original source
+        ctypes.memset(keypair_buf, 0, 64)
```
keypair_buf is 96 bytes, should clear them all

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/eac92da8cfb393e1704046103ac90e0b14fd2160

dgpv commented:
After starting to work on taproot support for python-elementstx, I added a bunch of changes here, related to the fact that Elements has different parameters for tagged hashes and has different default leaf version.

I moved hashers and default leaf version to `CoreChainParams`: https://github.com/Simplexum/python-bitcointx/commit/0db3f0ae826652171b4a61bcd6be28948535a6e2

Moved tap tweak-related code from methods of XOnlyPubKey to standalone functions: https://github.com/Simplexum/python-bitcointx/commit/d3174485dc81337843f8f5684bc719d7beef3686
This is needed because `XOnlyPubKey` is not a 'dispatched' class, that is, there's no `BitcoinXOnlyPubKey` or `ElementsXOnlyPubKey`, but the hashers are chain-dependent. With them as standalone functions, there will be no confusion on the context where they are used, while if they are methods, there might be confusion when `XOnlyPubKey` is created in Elements context and used in Bitcoin context, but the user would expect the tweak functions to behave as if in Elements context.

I split sign_schnorr into `sign_schnorr_no_tweak` and `sign_schnorr_tweaked`, where the later is only available from `CCoinKey`: https://github.com/Simplexum/python-bitcointx/commit/859d82cb28268f43ceec0f4e200e032bd9011175
This is also because the tweaks are chain-context-dependent, and `CKey` is not a 'dispatched' class, but `CCoinKey` _is_ dispatched, so the correct tags will be used inside `CCoinKey`'s methods when doing tagged hashing.

AdamISZ commented:
I was able to pay into a P2TR address (the plain keypath spend type) and spend out of it on regtest. It's a good sanity check to start. Obviously there is a certain amount of just "messing about" in here, but just for insight I've copied it out. More comments below.

```
(venv) $ python
Python 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import bitcointx as btc
>>> btc.allow_secp256k1_experimental_modules()
>>> btc.select_chain_params("bitcoin/regtest")
(<BitcoinMainnetParams: 'bitcoin'>, <BitcoinRegtestParams: 'bitcoin/regtest'>)
>>> from bitcointx.core.key import CKey
>>> priv = b"\x01"*32
>>> k = CKey(priv)
>>> k.xonly_pub
b'\x1b\x84\xc5V{\x12d@\x99]>\xd5\xaa\xba\x05e\xd7\x1e\x184`H\x19\xff\x9c\x17\xf5\xe9\xd5\xdd\x07\x8f'
>>> k.pub
CPubKey(x('031b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f'))
>>> from bitcointx.wallet import P2TRCoinAddress
>>> P2TRCoinAddress.from_xonly_pubkey(k.xonly_pub)
P2TRBitcoinRegtestAddress('bcrt1p33wm0auhr9kkahzd6l0kqj85af4cswn276hsxg6zpz85xe2r0y8s7hfsm7')
>>> P2TRCoinAddress.from_pubkey(k.pub)
P2TRBitcoinRegtestAddress('bcrt1p33wm0auhr9kkahzd6l0kqj85af4cswn276hsxg6zpz85xe2r0y8s7hfsm7')
>>> # here i used sendtoaddress in bitcoin-cli to fund the address.
>>> hextx1 = "0200000000010173a9d04d20c5d44d1bbbdb1abef2e6d053bb665f2da3cd31bced6b5eb06c07d00100000000feffffff0280969800000000002251208c5db7f797196d6edc4dd7df6048f4ea6b883a6af6af032342088f436543790f700fe40000000000160014780c6de5527e1462669e975d360f80a524a0833c024730440220661da93bd1a5f0095e89bc491285971d9f1c6c135ad7dedb1de4601b26c1c92c022039d8ed77749906ea5cc0f73cc526c358beec37df073dea03e07b9882850a52f2012102411ebe780264174df55a226d1bb41d3c6e79b06734842e1a04b4b7629bbb63c968030000"
>>> from binascii import hexlify, unhexlify
>>> tx1 = unhexlify(hextx1)
>>> from bitcointx.core import CTransaction
>>> CTransaction.deserialize(tx1)
CBitcoinTransaction([CBitcoinTxIn(CBitcoinOutPoint(lx('d0076cb05e6bedbc31cda32d5f66bb53d0e6f2be1adbbb1b4dd4c5204dd0a973'), 1), CBitcoinScript([]), 0xfffffffe)], [CBitcoinTxOut(0.1*COIN, CBitcoinScript([1, x('8c5db7f797196d6edc4dd7df6048f4ea6b883a6af6af032342088f436543790f')])), CBitcoinTxOut(0.1494616*COIN, CBitcoinScript([0, x('780c6de5527e1462669e975d360f80a524a0833c')]))], 872, 2, CBitcoinTxWitness([CBitcoinTxInWitness(CScriptWitness([x('30440220661da93bd1a5f0095e89bc491285971d9f1c6c135ad7dedb1de4601b26c1c92c022039d8ed77749906ea5cc0f73cc526c358beec37df073dea03e07b9882850a52f201'),x('02411ebe780264174df55a226d1bb41d3c6e79b06734842e1a04b4b7629bbb63c9')]))]))
>>> outs = [{"address": "", "value": 9998000}]
>>> priv2 = b"\x02"*32
>>> k2 = CKey(priv2)
>>> outaddr = P2TRCoinAddress.from_xonly_pubkey(k2.xonly_pub)
>>> outaddr
P2TRBitcoinRegtestAddress('bcrt1p5e6v9v2j5wp3y6c79gaqdqltq7jdv45fswnnm7exmmp2020mqepspf6x45')
>>> #this is the format we use in JM, but not needed here:
>>> outs = [{"address": str(outaddr), "value": 9998000}]
>>> outs
[{'address': 'bcrt1p5e6v9v2j5wp3y6c79gaqdqltq7jdv45fswnnm7exmmp2020mqepspf6x45', 'value': 9998000}]
>>> intxid = unhexlify("a8ab6b6740182bcec11e36a65f0b3baf88969ca3f32d24df2fe4f75875ed8fed")
>>> ins = [(intxid, 0)]
>>> from bitcointx.core import CMutableOutPoint
>>> from bitcointx.core import CMutableTxIn
>>> outpoint = CMutableOutPoint(intxid[::-1], 0)
>>> vin = [CMutableTxIn(prevout=outpoint, nSequence=0xffffffff)]
>>> vin
[CBitcoinMutableTxIn(CBitcoinMutableOutPoint(lx('a8ab6b6740182bcec11e36a65f0b3baf88969ca3f32d24df2fe4f75875ed8fed'), 0), CBitcoinScript([]), 0xffffffff)]
>>> from bitcointx.wallet import CCoinAddress
>>> sPK = CCoinAddress(outs[0]["address"]).to_scriptPubKey()
>>> sPK
CBitcoinScript([1, x('a674c2b152a383126b1e2a3a0683eb07a4d6568983a73dfb26dec2a7a9fb0643')])
>>> from bitcointx.core import CMutableTxOut
>>> vout = [CMutableTxOut(outs[0]["value"], sPK)]
>>> vout
[CBitcoinMutableTxOut(0.09998*COIN, CBitcoinScript([1, x('a674c2b152a383126b1e2a3a0683eb07a4d6568983a73dfb26dec2a7a9fb0643')]))]
>>> from bitcointx.core import CMutableTransaction
>>> tx2 = CMutableTransaction(vin, vout, nVersion=2)
>>> tx2
CBitcoinMutableTransaction([CBitcoinMutableTxIn(CBitcoinMutableOutPoint(lx('a8ab6b6740182bcec11e36a65f0b3baf88969ca3f32d24df2fe4f75875ed8fed'), 0), CBitcoinScript([]), 0xffffffff)], [CBitcoinMutableTxOut(0.09998*COIN, CBitcoinScript([1, x('a674c2b152a383126b1e2a3a0683eb07a4d6568983a73dfb26dec2a7a9fb0643')]))], 0, 2, CBitcoinMutableTxWitness([CBitcoinMutableTxInWitness(CScriptWitness([]))]))
>>> from bitcointx.core import CTxOut
>>> cto1 = CTxOut(10000000, P2TRCoinAddress.from_xonly_pubkey(k.xonly_pub).to_scriptPubKey())
>>> cto1
CBitcoinTxOut(0.1*COIN, CBitcoinScript([1, x('8c5db7f797196d6edc4dd7df6048f4ea6b883a6af6af032342088f436543790f')]))
>>> spent_outputs = [cto1]
>>> from bitcointx.core.script import SignatureHashSchnorr
>>> SignatureHashSchnorr(tx2, 0, spent_outputs)
b'\xae\x92\x97)\xd2\x84\xe3\x1a\xbd\xfcyC\xee\xc9y\x07v\xa7\x04s\x8d\xee\x96\xad\xa4\x03\x85!Sp\r\x92'
>>> sighash = SignatureHashSchnorr(tx2, 0, spent_outputs)
>>> sig = k.sign_schnorr_tweaked(sighash)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'CKey' object has no attribute 'sign_schnorr_tweaked'
>>> from bitcointx.wallet import CCoinKey
>>> k = CCoinKey.from_secret_bytes(b"\x01"*32)
>>> sig = k.sign_schnorr_tweaked(sighash)
>>> len(sig)
64
>>> from bitcointx.core import CMutableTxInWitness
>>> from bitcointx.core.script import CScriptWitness
>>> tx2.wit.vtxinwit[0] = CMutableTxInWitness(CScriptWitness([sig]))
>>> tx2
CBitcoinMutableTransaction([CBitcoinMutableTxIn(CBitcoinMutableOutPoint(lx('a8ab6b6740182bcec11e36a65f0b3baf88969ca3f32d24df2fe4f75875ed8fed'), 0), CBitcoinScript([]), 0xffffffff)], [CBitcoinMutableTxOut(0.09998*COIN, CBitcoinScript([1, x('a674c2b152a383126b1e2a3a0683eb07a4d6568983a73dfb26dec2a7a9fb0643')]))], 0, 2, CBitcoinMutableTxWitness([CBitcoinMutableTxInWitness(CScriptWitness([x('196127ee7b6c2d7bfd9863ed3e9b4e912878a0a80935fe72aab91936264486741524e8d04c16ea76e80f6f20541ff954ab129725d3e5024bfadb17afd4e5a71b')]))]))
>>> CTransaction.deserialize(tx1)
CBitcoinTransaction([CBitcoinTxIn(CBitcoinOutPoint(lx('d0076cb05e6bedbc31cda32d5f66bb53d0e6f2be1adbbb1b4dd4c5204dd0a973'), 1), CBitcoinScript([]), 0xfffffffe)], [CBitcoinTxOut(0.1*COIN, CBitcoinScript([1, x('8c5db7f797196d6edc4dd7df6048f4ea6b883a6af6af032342088f436543790f')])), CBitcoinTxOut(0.1494616*COIN, CBitcoinScript([0, x('780c6de5527e1462669e975d360f80a524a0833c')]))], 872, 2, CBitcoinTxWitness([CBitcoinTxInWitness(CScriptWitness([x('30440220661da93bd1a5f0095e89bc491285971d9f1c6c135ad7dedb1de4601b26c1c92c022039d8ed77749906ea5cc0f73cc526c358beec37df073dea03e07b9882850a52f201'),x('02411ebe780264174df55a226d1bb41d3c6e79b06734842e1a04b4b7629bbb63c9')]))]))
>>> hexlify(tx2.serialize())
b'02000000000101ed8fed7558f7e42fdf242df3a39c9688af3b0b5fa6361ec1ce2b1840676baba80000000000ffffffff01b08e980000000000225120a674c2b152a383126b1e2a3a0683eb07a4d6568983a73dfb26dec2a7a9fb06430140196127ee7b6c2d7bfd9863ed3e9b4e912878a0a80935fe72aab91936264486741524e8d04c16ea76e80f6f20541ff954ab129725d3e5024bfadb17afd4e5a71b00000000'
```

A few comments: Initially, I found myself rather confused about `merkle_root` argument, at least specifically for this no-script, provably-keypath-only spending case as recommended in the BIP and implemented here.

For example I think the docstring here still isn't quite enough for a user to be absolutely sure what it's doing:

```
"""Create a P2TR address from x-only "internal" pubkey (in BIP341 terms)

Raises CCoinAddressError if pubkey is invalid
"""
```
I think it should state that the actual recommendation for provably-no-script path (as justified in cite note 22) is being followed (or maybe just: that the default procedure `Q = P + H(P||m)G` is being followed with `m == b''`), because it isn't immediately obvious from code that you are calling to generate this address from a pubkey is using `compute_tap_tweak_hash` with `merkle_root == b''` (I mean, sure it's not *that* hard to figure it out, but I felt obliged to do so because the doc didn't give me 100% certainty what procedure is being followed).

Meanwhile I *do* think this docstring for `from_xonly_output_pubkey` is clear to a reader, and that should help a lot:

```
"""Create a P2TR address from x-only pubkey that is already tweaked,
        the "output pubkey" in the terms of BIP341

        Raises CCoinAddressError if pubkey is invalid, unless accept_invalid
        is True.
        """
```

Another aspect of this: that `_sign_schnorr_internal` uses `merkle_root is None` as the differentiator for not tweaking, while `merkle_root == b''` is different. I do realise that it's unambiguous and correct, but just mentioning that it seems like the kind of thing that can trip people up (like, they might think they should call `sign_schnorr_no_tweak` if they don't have a script path).

You can also see I originally didn't notice the `CCoinKey` part that you mentioned above. But that's clear now.

I think it was pretty smooth and intuitive apart from that, inasmuch as, it's very similar to the pre-existing workflow. But then of course I *only* tried the simplest possible version of the scriptpubkey to spend. At some point I will try using script path stuff, but it's less of a priority for me as I don't have any use case for now. All the same I will read the script part of the patch a bit more and add any further comments from what I see. Btw search `scnorr` and `shnorr`, I think you have a typo in one or two places.

dgpv commented:
https://github.com/Simplexum/python-bitcointx/pull/60/commits/af7e2de737e610348bb5d89e686fcc622d5035af

I've noticed that XOnlyPubKey did not have `__repr__` implemented, I fixed this.
Also I've searched for `scnorr` and `shnorr`, found two mistypes, fixed.
Added docstrings that talk about the differences between different tweak 'modes'.

> _sign_schnorr_internal uses merkle_root is None as the differentiator for not tweaking, while merkle_root == b'' is different. I do realise that it's unambiguous and correct, but just mentioning that it seems like the kind of thing that can trip people up (like, they might think they should call sign_schnorr_no_tweak if they don't have a script path)

because this is internal function, I felt that it is ok to use `None`/`b''`/`<32 bytes>` distinction instead of using special flag object as I did before. The function is not supposed to be called from outside the library code, so anyone who calls it is supposedly know that they are doing, as internal functions are not part of API. Still, I added docstring that hopefully gives explanations to those who are looking at the code.

AdamISZ commented:
Great. On that last point, of course, well understood that it is not for external callers; the reason it crops up is because I'm trying to read it carefully to understand exactly what happens when I call the API-level functions. And that's exactly what I *don't* need to do now because the docstrings are a lot more detailed :) (well, it's always interesting to read the code more, of course! but anyway).

dgpv commented:
(last force-push to https://github.com/Simplexum/python-bitcointx/commit/6270e628a6caa953ee74bcadcafd1afef8fa4bce was just to fix docstring of `sign_schnorr_tweaked()` to make it more readable)

dgpv commented:
(another force-push https://github.com/Simplexum/python-bitcointx/commit/35a53b5f4b41cc6c0cf37826f1774cd392c3b37b to fix a typo in docstring)

dgpv commented:
If I understand correctly, creating provably unspendable internal pubkey would go something like this:

```
shared_secret_key = CCoinKey.from_secret_bytes(os.urandom(32))
nums_tweak = lx('0250929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0')
no_keypath_internal_pub = tap_tweak_pubkey(shared_secret_key.xonly_pub, nums_tweak)
```

It is then easy to use `no_keypath_internal_pub` with `TaprootScriptTree` in a normal way.

I don't think we should hide the nums point inside some method of `TaprootScriptTree`, or export it as some constant. This is just one possible NUMS, 'the one featured in BIP341', but we cannot even say that it is a 'recommended' one, because the BIP gives it just as an example. I think the choice of NUMS point is better left to the projects that use the library rather than the library itself.

dgpv commented:
Hmm, maybe it is a good idea to rename `merkle_root` argnument of `tap_tweak_pubkey` to just `tweak`, because in this usage `merkle_root` reference is not applicable...

dgpv commented:
Hmm, no, on the second thought, it is not a good idea to rename `merkle_root` into `tweak`, because `merkle_root` is used further in `compute_tap_tweak_hash`, and `merkle_root` is the most appropriate name there. And using `tap_tweak_pubkey` is not the only way to create the 'unspendable keypath internal key'. One could do

```
no_keypath_internal_pub = XOnlyPubKey(CPubKey.add(shared_secret_key.pub, nums_pub))
```

It will result in a different pubkey, of course, because there's no tagged hashing involved, but the overall meaning would be the same, IIUC.

AdamISZ commented:
Do I understand correctly that I should use `CKeyBase.sign_schnorr_no_tweak` for signatures inside script path execution (e.g. added to witness as elements to be consumed in a script path spend of, say, a multisig with OP_CHECKSIGADD)? I'm having a bit of trouble finding examples of this to confirm my understanding.

dgpv commented:
> Do I understand correctly that I should use CKeyBase.sign_schnorr_no_tweak for signatures inside script path execution 

Yes.

AdamISZ commented:
I recorded a successful signet spend on a script path spending [here](https://x0f.org/web/statuses/107322448218791237); if I made any errors in my description please let me know, don't want to mislead readers :)

dgpv commented:
> it arranges your list of `CScript` objects into a merkle tree (balanced; nothing fancy)

You can make unbalanced trees - you just use another `TaprootScriptTree` as a leaf (the docstring of `TaprootScriptTree` class actually shows an example of unbalanced tree)

AdamISZ commented:
> You can make unbalanced trees - you just use another TaprootScriptTree as a leaf (the docstring of TaprootScriptTree class actually shows an example of unbalanced tree)

Oh yes, I was just thinking that: it's neat that you can nest trees. Indeed my comment there is a bit off, for that reason.

dgpv commented:
Sorry for slipping an unrelated fix for signet adr prefixes into this PR. As a (lame) excuse I'll say that the problem with signet prefixes were found during testing of the taproot spends on signet, and I wanted this fix to be in the taproot branch :-)

dgpv commented:
I've added another PR https://github.com/Simplexum/python-bitcointx/pull/62 that is on top of this one. That PR migrates automatic testing to github actions, and also lists the 'recommended' libsecp256k1 commit in the README, with extensive comments on dynamic linking and its ABI instability

dgpv commented:
Moved the fix for signet addrs an keys to master, as the fix is trivial and is not related to this PR: https://github.com/Simplexum/python-bitcointx/commit/50d4e94de0186d3670371eff104f74a047e01e81

dgpv commented:
Moved two other unrelated small changes to master: https://github.com/Simplexum/python-bitcointx/commit/a4b5fd890fa520213fae2800b2147074bd44b52d https://github.com/Simplexum/python-bitcointx/commit/43e4415ce3a15c8c589fe0aee6960f521e2c9b85

Squashed the taproot changes.

dgpv commented:
(another rebase)

dgpv commented:
Excerpt from the discussion on Bitcoin Core slack, comments by @roconnor-blockstream:

> Russell O'Connor:
> So the taproot signatures commit to the tapleaf, but not to the merkle path.
> The problem is that the same tapleaf could occur multiple times in the same merkle tree.
> in particular the same tapleaf can occur at different depths.
> Thus if a tapleaf does occur at different depths, then a second-party can maleate your transaction to move its redemption from one tapleaf to another copy, in particular at a deeper depth.
> This doesn't change the functional properties of the transaction, but it does increase the weight of your transaction, and thus lowers your intended fee rate.
> The work around is to never participate in a taproot address where you do not know all the tapleaves, and in particular make sure none of the tapleaves are duplicated.
> Which is probably good advice regardless.

I wonder if it is sensible for `TaprootScriptTree(leaves_or_branches)` to raise `ValueError` by default if `leaves_or_branches` contain duplicate leaves (and allow such duplicates if a special flag `allow_duplicate_leaves=True` is supplied)

AdamISZ commented:
> I wonder if it is sensible for TaprootScriptTree(leaves_or_branches) to raise ValueError by default if leaves_or_branches contain duplicate leaves (and allow such duplicates if a special flag allow_duplicate_leaves=True is supplied)

Interesting.

For what it's worth, my opinion: I don't think this library needs to do that. If there are indeed use cases for duplicate leaves at different points in the tree, then it'd be up to the people creating, reading, signing such transactions (i.e. one step up the software stack from this library) to decide how to deal with it. I suppose my argument is: there are an infinite number of ways in which a leaf might be dangerous to a user, and this is only one; clearly if a user doesn't know the full tree, then there might be a spending path that they should not accept (so agreeing with roconnor's "probably good advice regardless").

dgpv commented:
(rebased on current master)

dgpv commented on bitcointx/core/key.py:
```diff
@@ -1905,6 +2073,178 @@ def get_pubkey(self, key_id: bytes,
+
+        result = _secp256k1.secp256k1_schnorrsig_verify(
+            secp256k1_context_verify,
+            sigbytes, hash, 32, self._to_ctypes_char_array()
```
We should check for `self.is_fullyvalid()` before this, and return `False` on invalid pubkey. Otherwise  `_to_ctypes_char_array()` will crash on assert

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698

dgpv commented on bitcointx/core/key.py:
```diff
@@ -1905,6 +2073,178 @@ def get_pubkey(self, key_id: bytes,
+
+    result = _secp256k1.secp256k1_xonly_pubkey_tweak_add_check(
+        secp256k1_context_verify, tweaked_pub, int(bool(parity)),
+        internal_pub._to_ctypes_char_array(), tweak)
```
need to check `internal_pub.is_fullyvalid()` to avoid crashing on assert inside `_to_ctypes_char_array()`

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698

dgpv commented on bitcointx/core/script.py:
```diff
@@ -1364,6 +1437,128 @@ def SignatureHash(script: CScript, txTo: 'bitcointx.core.CTransaction', inIdx: i
+    tapleaf_hash: Optional[bytes] = None,
+    codeseparator_pos: int = -1,
+    annex_hash: Optional[bytes] = None
+) -> bytes:
```
`tapleaf_hash`, `annex_hash`, and `SIGVERSION_TAPSCRIPT` are currently not covered by tests

AdamISZ commented on bitcointx/core/key.py:
```diff
@@ -1905,6 +2073,178 @@ def get_pubkey(self, key_id: bytes,
+def tap_tweak_pubkey(pub: XOnlyPubKey, *, merkle_root: bytes = b'',
+                     ) -> Optional[Tuple[XOnlyPubKey, bool]]:
+
+    base_point = pub._to_ctypes_char_array()
```
Presumably the same issue as above, https://github.com/Simplexum/python-bitcointx/pull/60/files/7dbad630d413758efb899cabd597d7f9fe903c14#r795157000 ?

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698

dgpv commented on bitcointx/tests/test_scripteval.py:
```diff
@@ -174,38 +313,56 @@ def test_script(self) -> None:
+
+    @unittest.skipIf(
+        not bitcointx.util._allow_secp256k1_experimental_modules,
+        "bitcoinconsensus library is not available or "
```
message is not correct here

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698

dgpv commented on bitcointx/wallet.py:
```diff
@@ -476,6 +484,160 @@ def from_redeemScript(cls: Type[T_P2WPKHCoinAddress],
+        tt_res = tap_tweak_pubkey(pubkey)
+
+        if not tt_res:
+            raise ValueError('cannot create tap tweak from supplied pubkey')
```
The convention before was that methods of address classes raise instances of `CCoinAddressError`, as for example above, where `is_fullyvalid()` is checked. I believe this is not the best state of things, as raising `ValueError` in `from_*` methods when the argument (pubkey for example) is somehow bad, seems more logical to me. But doing so means that we will be breaking an established API. `CCoinAddressError` is direcltly descended from `Exception`, not from `ValueError`. Existing code would expect `CCoinAddressError`, not `ValueError` when calling `from_*` methods.

Therefore I think  we should raise `P2TRCoinAddressError` here.

dgpv replied:
We could make `CCoinAddressError` a descendant of `ValueError`. It would still constitute some API breakage, but I believe it would be much lesser breakage. We could say that an incorrect encoding of the address string is a 'value error`, after all - the value of the string as supplied to `CCoinAddress.__new__()`

dgpv replied:
The breakage in this case would be that the code that catches `ValueError` will begin to catch `CCoinAddressError` and descendants, too.

AdamISZ replied:
On this topic (whether exceptions raised are only `*CoinAddressError` type), isn't it the case that this line can also raise `ValueError` with a wrong data length?

https://github.com/Simplexum/python-bitcointx/blob/fd1af3d04c75ccfa96c1e6940da480867ab5d736/bitcointx/wallet.py#L533

AdamISZ replied:
Fwiw, I agree with the line of thinking that it's better that all of these raises are `*CoinAddressError` type.

dgpv replied:
>  isn't it the case that this line can also raise ValueError with a wrong data length?

Yes, this is the case. If we're to raise `CCoinAddressError`, we need to catch `ValueError` here

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698

dgpv commented on bitcointx/wallet.py:
```diff
@@ -476,6 +484,160 @@ def from_redeemScript(cls: Type[T_P2WPKHCoinAddress],
+        if not stree.internal_pubkey:
+            raise ValueError(
+                f'The supplied instance of {stree.__class__.__name__} '
+                f'does not have internal_pubkey')
```
The same issue of `ValueError` vs `CoinAddressError` as in https://github.com/Simplexum/python-bitcointx/pull/60/files#r795441715

dgpv replied:
Fixed in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698

dgpv commented on bitcointx/wallet.py:
```diff
@@ -476,6 +484,160 @@ def from_redeemScript(cls: Type[T_P2WPKHCoinAddress],
+        form.
+        """
+        if scriptPubKey.is_witness_v1_taproot():
+            return cls.from_bytes(scriptPubKey[2:34])
```
We could check validity of the included x-only pubkey here and raise an error if it is invalid, but I'm not sure if that would be actually helpful

dgpv replied:
I decided that checking for validity of pubkey in the scriptPubKey is not what the library should do. But I added the text to the docstring in https://github.com/Simplexum/python-bitcointx/commit/49cb1613cb66e556e7d4deb669717e7de6bf6698 that tells that the validity of pubkey is _not_ checked.

dgpv commented:
Synced the taproot branch with https://github.com/Simplexum/python-bitcointx/tree/segwitv1_addrs

dgpv commented:
rebased on current master

dgpv commented:
merged

HRezaei commented:
Thank you @dgpv for all your efforts.
> Please don't hesitate to comment if you think something might be made more convenient for the developer. 

The most crucial thing needed now, is some examples, I would say. Even very simple, but fully working when sent to network.

I tried one for keypath spending similar to [this comment](https://github.com/Simplexum/python-bitcointx/pull/60#issuecomment-974832099) by @AdamISZ. It produces a transaction hex, but when sent to a network, an error is returned like this:
`bitcoin.rpc.JSONRPCError: {'code': -344, 'message': 'Error validating transaction: Transaction missing input or output..'}`

My code is attached below, any comments to resolve the issue would be appreciated. I think this, once resolved, can be added to examples:

```python
import bitcointx as btc
btc.allow_secp256k1_experimental_modules()

from binascii import unhexlify
from bitcointx.core import CTransaction
from bitcointx.wallet import P2TRBitcoinTestnetAddress
from bitcointx.core import CMutableOutPoint
from bitcointx.core import CMutableTxIn
from bitcointx.wallet import CCoinAddress
from bitcointx.core import CMutableTxOut
from bitcointx.core import CMutableTransaction
from bitcointx.core import CTxOut
from bitcointx.core.script import SignatureHashSchnorr
from bitcointx.wallet import CCoinKey
from bitcointx.core import CMutableTxInWitness
from bitcointx.core.script import CScriptWitness
from bitcoin.rpc import Proxy


def parse_raw_tx(tx_hex):
    parsed_tx = {
        'version': tx_hex[:8],
        'Input Count': tx_hex[8:10],
        'Input 1 Previous Output Hash': tx_hex[10:74],
    }
    return parsed_tx


print(btc.select_chain_params("bitcoin/testnet"))


def test_tx_from_raw_byte_keys():
    priv = b"\x01"*32
    alice_priv_key = CCoinKey.from_secret_bytes(priv)
    print(alice_priv_key.xonly_pub)

    print(alice_priv_key.pub)

    alice_address = P2TRBitcoinTestnetAddress.from_xonly_pubkey(alice_priv_key.xonly_pub)
    print("Alice address: ", alice_address)

    print(P2TRBitcoinTestnetAddress.from_pubkey(alice_priv_key.pub))

    # here i used a faucet to fund the address.
    tx_id = "86e68fc34258372e96626841c3b05ae6b478d20fe85a769bca91bf1b13f8c83c"
    prev_tx_hex = "02000000000101bdf4080aaefe9d274f5f91ca1250d81cd05e62d67641bbf52e71e246ee8557e40000000000feffffff02abbb0f0000000000160014c885ac9ecdee154b1ac3a372ebc1184601b9da20204e0000000000002251208c5db7f797196d6edc4dd7df6048f4ea6b883a6af6af032342088f436543790f0247304402205de0c18eab95151c1a2908b247c4c5197894a4587ed0326cba06976183f7b058022067251f579e85f027368056cd1fda0a0baef9a5dac69ee6736e2d84f95ee3fa410121034bcc56646b1e8f619522c3db1feebd64bac532a5e09389e4805caf55cf38d3da167b2100"
    prev_tx_bytes = unhexlify(prev_tx_hex)
    prev_tx = CTransaction.deserialize(prev_tx_bytes)

    print("Previous tx: ", prev_tx)
    print("Previous tx parsed: ", parse_raw_tx(prev_tx_hex))
    print("Available balance: ", prev_tx.vout[1].nValue)
    output_index_in_prev_tx = 1
    balance = prev_tx.vout[output_index_in_prev_tx].nValue
    fee = 2000
    amount = balance - fee

    priv2 = b"\x02"*32
    bob_priv_key = CCoinKey.from_secret_bytes(priv2)
    bob_addr = P2TRBitcoinTestnetAddress.from_xonly_pubkey(bob_priv_key.xonly_pub)
    print("Bob address: ", bob_addr)

    # this is the format we use in JM, but not needed here:
    outs = [{"address": str(bob_addr), "value": amount}]
    print("outs:", outs)

    intxid = unhexlify(tx_id)
    outpoint = CMutableOutPoint(intxid[::-1], output_index_in_prev_tx)
    vin = [CMutableTxIn(prevout=outpoint, nSequence=0xffffffff)]
    print("vin:", vin)

    sPK = CCoinAddress(outs[0]["address"]).to_scriptPubKey()
    print("sPK:", sPK)

    vout = [CMutableTxOut(outs[0]["value"], sPK)]
    print("vout:", vout)

    tx = CMutableTransaction(vin, vout, nVersion=2)

    cto1 = CTxOut(balance, alice_address.to_scriptPubKey())
    print("cto1:", cto1)

    spent_outputs = [cto1]
    print("SignatureHashSchnorr:", SignatureHashSchnorr(tx, 0, spent_outputs))

    sighash = SignatureHashSchnorr(tx, 0, spent_outputs)

    sig = alice_priv_key.sign_schnorr_tweaked(sighash)
    print(len(sig))

    tx.wit.vtxinwit[0] = CMutableTxInWitness(CScriptWitness([sig]))
    print("tx:", tx)

    connection = Proxy(service_url='http://api.blockcypher.com/v1/btc/test3/txs/push')
    res = connection.sendrawtransaction(tx)
    print(res)


if __name__ == '__main__':
    test_tx_from_raw_byte_keys()


```

AdamISZ commented:
@HRezaei just in case it helps, I'm guessing you didn't see this example I did around that time, which tested script path and which broadcast OK: https://gist.github.com/AdamISZ/6233d9d9b8d25483dc5d39cc6b9892a7

HRezaei commented:
> @HRezaei just in case it helps, I'm guessing you didn't see this example I did around that time, which tested script path and which broadcast OK: https://gist.github.com/AdamISZ/6233d9d9b8d25483dc5d39cc6b9892a7

Thanks @AdamISZ,
But I just tried to broadcast that tx in your example as well and it raises the same error. Maybe my way of broadcasting is wrong?! 🤔
```python
from bitcoin.rpc import Proxy
connection = Proxy(service_url='http://api.blockcypher.com/v1/btc/test3/txs/push')
res = connection.sendrawtransaction(tx2)
print(res)
```

AdamISZ commented:
> But I just tried to broadcast that tx in your example as well and it raises the same error.

Presumably you don't mean that literally? That transaction of course wouldn't broadcast, it's already been mined a month or two back :) But in that case I wonder what you do mean?

Re: using a proxy, I'd just use a signet or testnet (highly prefer signet! testnet3 is awful for many reasons) Core instance and use `sendrawtransaction`. Less chance of something going wrong with the intermediary and clearer error messages.

HRezaei commented:
You're right @AdamISZ, 
I guess it's better to continue the discussion under your own gist...
